### PR TITLE
feat: env vars should be readonly

### LINKED ISSUE #108: environment variables should be read-only, but they are not
Consider the following code:

```ts
export const env = createEnv({
  server: {
    PORT: z.number().default(80),
  }
});

env.PORT = 123; // Works??? WTF
```

Assignment here should throw an error from the TypeScript compiler, since we don't want to modify environment variables at runtime. If mutations are actually required, then they should be done using actual new TypeScript variables. (This is basically the same rationale as preferring to use `const` to make new variables over `let` if possible. In fact, by naming the properties of the object as uppercase, you are violating the mutability contract that TypeScript users expect when seeing uppercase variables/properties.)

I think submitting a PR for this should be pretty straightforward if you want me to do that, we just need to use the `Readonly` utility type.

### PR DESCRIPTION
**Pull request by Zamiell**
Closes #108

### CODE REVIEW

File: examples/astro/src/env.d.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+/// <reference types="astro/client" />
```
Comment (Zamiell) [inline examples/astro/src/env.d.ts:1]: looks like this file got created somehow when i ran `pnpm run test` or `pnpm run lint` or `pnpm run typecheck`.

should it be added to .gitignore maybe?

Comment (juliusmarminge) [reply examples/astro/src/env.d.ts:1]: Ye probably - I've not Skipper checking it in 😅

**Review by juliusmarminge — APPROVED** [review]
Add a test like this too for good measure and to show what we're after:

```ts
// @ts-expect-error - readonly
env.FOO = true
```

File: packages/core/test/smoke.test.ts [hunk 2]
```diff
@@ -369,3 +379,15 @@ describe("shared can be accessed on both server and client", () => {
...
+test("envs are readonly", () => {
+  const env = createEnv({
+    server: { BAR: z.string() },
+    runtimeEnv: { BAR: "bar" },
+  });
+
+  // @ts-expect-error - envs are readonly
+  env.BAR = "foo";
+
+  expect(env).toMatchObject({ BAR: "bar" });
```
Comment (juliusmarminge) [inline packages/core/test/smoke.test.ts:392]: this currently fails

File: .gitignore [hunk 3]
```diff
@@ -9,5 +9,6 @@ dist
 .vercel
 .DS_Store
 
+env.d.ts
```
Comment (Zamiell) [inline .gitignore:12]: this is my fix for the aforementioned gitignore issue.
please double check this before merging.

**Comment by Zamiell** [comment]
> Should we also Object.freeze something?

Personally I'm not a fan of `Object.freeze`.

### Pros

- Ensures that the readonly state persists into the runtime, making the object even safer.

### Cons

- `Object.freeze` is superfluous in TypeScript-only applications. And we don't ever expect that someone would be exporting their application's environment variables to some kind of JavaScript-only consumer, so the runtime safety is truly pointless.
- Prevents live runtime debugging using the Chrome development console.

Overall it seems like the cons outweigh the pros. Also note that there is [no performance benefit to doing this](https://stackoverflow.com/questions/8435080/any-performance-benefit-to-locking-down-javascript-objects).

File: packages/core/test/smoke.test.ts (hunk 2 above)
Comment (Zamiell) [reply packages/core/test/smoke.test.ts:392]: the test failing is intentional because TypeScript doesn't exist at runtime

Comment (Zamiell) [reply packages/core/test/smoke.test.ts:396]: i'm not sure what the point of this test is.
the other tests that i modified already ensure that the type is readonly, so i think we can delete this test.

File: packages/core/index.ts [hunk 4]
```diff
@@ -209,6 +211,14 @@ export function createEnv<
...
       return target[prop as keyof typeof target];
     },
+    set(_target, prop) {
+      // Readonly - this is the error message you get from assigning to a frozen object
+      throw new Error(
+        typeof prop === "string"
+          ? `Cannot assign to read only property ${prop} of object #<Object>`
+          : `Cannot assign to read only property of object #<Object>`
+      );
+    },
```
Comment (juliusmarminge) [inline packages/core/index.ts:221]: what do you think about this?

File: packages/core/test/smoke.test.ts (hunk 2 above)
Comment (juliusmarminge) [reply packages/core/test/smoke.test.ts:396]: fixed it by making the `set` handler throw

File: packages/core/index.ts (hunk 4 above)
Comment (Zamiell) [reply packages/core/index.ts:221]: i dont like it for the reasons that I outlined in my previous post

File: packages/core/test/smoke.test.ts (hunk 2 above)
Comment (Zamiell) [reply packages/core/test/smoke.test.ts:396]: i'm against this test and this `set` handler for the reasons i outlined in my previous post

File: packages/core/index.ts (hunk 4 above)
Comment (juliusmarminge) [reply packages/core/index.ts]: reverted change

**Review by juliusmarminge — APPROVED** [review]

**Comment by Zamiell** [comment]
im not a fan of committing commented out code into a repository, can I remove this please?

**Comment by juliusmarminge** [comment]
> im not a fan of committing commented out code into a repository, can I remove this please?

Nah I want it as a reference to why the behavior is the way it is if we come back to it in the future

**Pull request merged**